File: leet_code_75/two_sum.js
// time: O(n)
//space: O(n)


let twoSum = function (nums, target) {
    let map = {}
    for (let i = 0; i < nums.length; i++) {

        let match = target - nums[i]

        if (match in map) return [i, map[match]]

        map[nums[i]] = i

    }
    return null
};

// let twoSum = (nums, target) => {
//     let map = {};

//     for (let i = 0; i < nums.length; i++) {
//         let another = target - nums[i];

//         if (another in map) {
//             return [map[another], i];
//         }

//         map[nums[i]] = i;
//     }

//     return null;
// };

File: algo_expert/arrays/two_sum.py
# O(n) time and space


def twoNumberSum(array, targetSum):
    matches = set()

    for num in array:
        match = targetSum - num
        if match in matches:
            return [num, match]
        else:
            matches.add(num)

    return []

# O(nlogn) time, O(1) space


def twoNumberSum1(array, targetSum):
	array.sort()
	left, right = 0, len(array) - 1

	while left < right:
		current_sum = array[left] + array[right]
		if current_sum == targetSum:
			return [array[left], array[right]]
		elif current_sum < targetSum:
			left += 1
		else:
			right -= 1

	return []
